src/Data/ListLike/Base.hs

```diff
@@ -104,34 +109,30 @@ class (IsList full, item ~ Item full, FoldableLL full item, Monoid full) =>
 
     {- | Extracts the first element of a 'ListLike'. -}
     head :: full -> item
-    head = maybe (error "head") fst . uncons
+    head = foldr (\ item _ -> item) (error "head")
 
     {- | Extract head and tail, return Nothing if empty -}
     uncons :: full -> Maybe (item, full)
     uncons x = if null x then Nothing else Just (head x, tail x) -- please don't
 
     {- | Extracts the last element of a 'ListLike'. -}
     last :: full -> item
-    last l = case genericLength l of
-                  (0::Integer) -> error "Called last on empty list"
-                  1 -> head l
-                  _ -> last (tail l)
+    last = foldl (\ _ item -> item) (error "Called last on empty list")
 
     {- | Gives all elements after the head. -}
     tail :: full -> full
     tail = maybe (error "tail") snd . uncons
 
     {- | All elements of the list except the last one.  See also 'inits'. -}
     init :: full -> full
-    init l
-        | null l = error "init: empty list"
-        | null xs = empty
-        | otherwise = cons (head l) (init xs)
-        where xs = tail l
+    init l =
+        case uncons l of
+            Just (x, xs) | not (null xs) -> cons x (init xs)
```

Comment:
Better, but is still not very elegant as it always does a lookahead of 2.  
To get lookahead of 1, implement it via `init1 :: item -> full -> full`, see, e.g. https://github.com/agda/agda/blob/a1231ba17621fa3334f229c8273af6c4e98ed68a/src/full/Agda/Utils/List.hs#L132-L138